### PR TITLE
ROX-21616: Use NVD CVSS scores enrichments and OSV CVSS scores

### PR DESCRIPTION
**Pull request by jvdm**
## Description

1. Pull CVSS scores from OSV updaters presented in `Severity`.
2. Use CVSS enrichments to populate CVSS

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO I have scanning results with CVSS scores run locally, need to run this in a cluster and post results here.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: scanner/mappers/mappers.go [hunk 1]
```diff
@@ -556,3 +466,170 @@ func convertSliceToMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT,
...
+	return out
+}
+
+// fixedInVersion returns the fixed in string, typically provided the report's
+// `FixedInVersion` as a plain string, but, in some OSV updaters, it can be an
+// urlencoded string or expected to be extracted from Range.Upper.
+func fixedInVersion(v *claircore.Vulnerability) string {
+	fixedIn := v.FixedInVersion
+	// Instead of matching the updater name, we simply attempt to retrieve from the
+	// severity.
```
[the remark this answers is not in the document]
Comment (jvdm) [reply scanner/mappers/mappers.go:485]: 4c4003125b

File: scanner/mappers/mappers.go [hunk 2]
```diff
@@ -556,3 +466,170 @@ func convertSliceToMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT,
...
+		}
+	}
+	return fixedIn
+}
+
+// nvdScoresMap look for NVD CVSS in the vulnerability report enrichments and
+// returns a mapping of vulnerability IDs to CVSS Scores V3.
+func nvdScoresMap(enrichments map[string][]json.RawMessage) (map[string]nvdschema.CVSSV30, error) {
+	cvss := enrichments[cvss.Type]
+	if len(cvss) <= 0 {
```
[the remark this answers is not in the document]
Comment (jvdm) [reply scanner/mappers/mappers.go:502]: 2dcd16e28

File: scanner/mappers/mappers.go [hunk 3]
```diff
@@ -556,3 +466,170 @@ func convertSliceToMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT,
...
+	err := json.Unmarshal(cvss[0], &scores)
+	if err != nil {
+		return nil, err
+	}
+	if len(scores) <= 0 {
+		return nil, nil
+	}
+	ret := make(map[string]nvdschema.CVSSV30)
+	for id, l := range scores {
+		// There not criteria for selecting more than one enrichment record, assume the
```
[the remark this answers is not in the document]
Comment (jvdm) [reply scanner/mappers/mappers.go:518]: 67704d4c9

File: scanner/mappers/mappers.go [hunk 4]
```diff
@@ -556,3 +466,170 @@ func convertSliceToMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT,
...
+	}
+	// TODO(ROX-21264): ClairCore only supports CVSS v3.1, so we assume CVSSV30 --
+	//                  need to update this when we start using our own CVSS score enricher.
+	var scores map[string][]nvdschema.CVSSV30
+	// The CVSS enrichment always contains only one element.
+	err := json.Unmarshal(cvss[0], &scores)
+	if err != nil {
+		return nil, err
+	}
+	if len(scores) <= 0 {
```
[the remark this answers is not in the document]
Comment (jvdm) [reply scanner/mappers/mappers.go:513]: 2dcd16e28

File: scanner/mappers/mappers.go [hunk 5]
```diff
@@ -556,3 +466,170 @@ func convertSliceToMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT,
...
+func severityAndScores(vuln *claircore.Vulnerability, nvdScores map[string]nvdschema.CVSSV30) (severityValues, error) {
+	var values severityValues
+	errList := errorhelpers.NewErrorList("parsing severity and CVSS scores")
+	switch {
+	case osvUpdaterPattern.MatchString(vuln.Updater):
+		if vuln.Severity == "" {
+			errList.AddStrings("severity is empty")
+			break
+		}
+		// ClairCore has no CVSS version indicator for OSV data, assuming CVSS V2 if
```
[the remark this answers is not in the document]
Comment (jvdm) [reply scanner/mappers/mappers.go:553]: 2a7e9f223

File: scanner/mappers/mappers.go [hunk 6]
```diff
@@ -556,3 +466,170 @@ func convertSliceToMap[IN any, OUT any](in []*IN, convF func(*IN) (string, *OUT,
...
+			}
+		}
+	default:
+		// Everything else uses NVD.
+		values.severity = vuln.Severity
+		if nvd, ok := nvdScores[vuln.ID]; ok {
+			switch {
+			case strings.HasPrefix(nvd.Version, "2"):
+				values.v3Score = float32(nvd.BaseScore)
+				values.v3Vector = nvd.VectorString
```
[the remark this answers is not in the document]
Comment (jvdm) [reply scanner/mappers/mappers.go:608]: Oh, a nice one. Thanks. 78cdfdc1d

**Review by RTann — APPROVED** [review]

**Pull request merged**